Stats: Fix date navigation arrows for sites using GMT offset

Nikschavan opened the pull request:
## Summary

Fixes the date navigation arrows in Jetpack Stats (Odyssey) not working correctly when the site timezone is set to a GMT offset (e.g., `gmt_offset: -8` for America/Los_Angeles) instead of an IANA timezone string.

### The Problem

When navigating dates using the arrow buttons:
1. The URL and date picker would show the correct date, but the title displayed a date off by one day
2. After clicking "Previous", clicking "Next" would do nothing (appeared disabled)

### Root Cause

In `getMomentSiteZone()`, when falling back to GMT offset handling:

1. **Date strings were parsed incorrectly**: `moment('2025-11-25')` parses as midnight in the browser's local timezone, then `.utcOffset(-8)` shifts the time, which could cross midnight and change the date
2. **Inconsistent offset handling**: The "today" moment (no dateInput) and parsed date moments had different offset settings, causing `isSameOrAfter()` comparisons to return incorrect results for enabling/disabling navigation arrows

### The Fix

Both code paths now use `moment.parseZone().utcOffset(gmtOffset, true)` where the second argument `true` is the `keepLocalTime` flag. This:
- Preserves the date/time values exactly as parsed
- Only changes the UTC offset metadata
- Ensures consistent comparisons between dates


https://github.com/user-attachments/assets/e14d1aba-3347-4d44-b7d3-39c9fa3c4531



## Related issues

Fixes STATS-186

## Testing instructions

### Pre-Requisite
1. Follow the testing instructions to test Odyssey stats - PCYsg-Pp7-p2

### Steps
1. Set up a site with timezone set to a non-UTC timezone (e.g., America/Los_Angeles which has `gmt_offset: -8`)
2. Open Jetpack Stats for a given day
3. Click the Previous arrow to navigate to the prior day
4. Verify the date shown in the title matches the URL and date picker
5. Click the Next arrow to navigate forward
6. Verify navigation works correctly and the Next arrow is not incorrectly disabled

Nikschavan committed:
Stats: Fix date navigation arrows for sites using GMT offset

When a site uses a GMT offset (e.g., gmt_offset: -8) instead of an IANA
timezone string, the date navigation arrows would break because:

1. Date strings were parsed in browser's local timezone then offset applied,
   which could shift the date across midnight boundaries
2. The "today" moment and parsed date moments had inconsistent offset handling,
   causing isSameOrAfter() comparisons to return incorrect results

This fix ensures both code paths use moment.parseZone().utcOffset(gmtOffset, true)
where the `true` flag (keepLocalTime) preserves the date values while only
changing the offset metadata. This makes date comparisons work correctly.

Fixes STATS-186

dognose24 reviewed (approved):
@Nikschavan  Thanks for the fix, the code changes work as described and LGTM! 👍🏼

|Before|After|
|-|-|
|<img width="869" height="712" alt="截圖 2025-11-26 晚上11 02 36" src="https://github.com/user-attachments/assets/24483c49-dfad-404a-82ba-b5b84b0a50b0" />|<img width="869" height="707" alt="截圖 2025-11-26 晚上11 03 08" src="https://github.com/user-attachments/assets/319e174a-940c-4a18-9ff1-8cf13ae52989" />|

kangzj committed:
Stats: Add comprehensive unit tests for useMomentInSite hook

Add test coverage for the useMomentInSite hook and getMomentSiteZone utility function, covering:
- IANA timezone string handling with validation
- DST transitions and seasonal offset changes
- GMT offset fallback behavior
- Locale handling and custom date formats
- Browser timezone fallback
- Hook memoization
- Multiple site ID scenarios

kangzj committed:
Stats: Simplify useMomentInSite hook implementation

Remove the unused dateFormat parameter from getMomentSiteZone and simplify the logic for handling undefined dateInput. The previous implementation was unnecessarily complex when creating moments for the current time with GMT offset.

Changes:
- Remove dateFormat parameter from getMomentSiteZone selector
- Simplify undefined dateInput handling in all branches
- Fix comparison logic in site.jsx to properly compare moment objects
- Remove obsolete custom date format tests

This refactoring maintains the same functionality while making the code more maintainable and easier to understand.

kangzj committed:
Stats: Refactor useMomentInSite to handle timezone semantics correctly

Improved the hook to properly distinguish between naive date strings and absolute times, ensuring correct timezone handling for different input types.

Key improvements:
- Accepts moment.MomentInput (strings, Date objects, moment instances) instead of only strings
- Naive date strings ('2024-01-15') are interpreted AS the site's timezone
- Date objects and ISO strings with timezone are converted TO the site's timezone
- Added regex pattern to detect timezone indicators (Z, +HH:MM, -HH:MM)
- Simplified code with better comments explaining each case

Added comprehensive test coverage (18 new tests):
- Tests for IANA timezone handling with various input types
- Tests for GMT offset handling with naive vs absolute times
- Edge cases: fractional offsets, zero offset, ISO format variations
- All 44 tests passing

This ensures dates like '2024-01-15' are treated as midnight in the site's timezone, while Date objects preserve absolute time and convert to the site's timezone correctly.

kangzj committed:
Stats: Fix realtime chart timezone handling

Fixed an issue where the realtime chart was showing incorrect time labels (e.g., -1283m) due to timezone mismatches.

The problem:
- Chart was using moment() to parse naive date strings like '2024-07-15 10:30:00'
- These strings were being parsed in browser's timezone instead of site's timezone
- When converted to Date objects via new Date(), the timestamps were incorrect
- This caused massive time differences in the formatTimeTick function

The fix:
- Use momentInSite() to parse all date strings in site's timezone
- Convert to Date objects using .toDate() method to preserve correct absolute time
- Ensures naive date strings are interpreted AS the site's timezone
- Removed unused moment import

Changes:
- Line 76: moment(eachMinute) → momentInSite(eachMinute)
- Line 89-90: Use momentInSite for hour comparisons
- Line 104: new Date(eachMinute) → momentInSite(eachMinute).toDate()

kangzj committed:
Revert realtime chart to use moment instead of momentInSite

kangzj committed:
Stats: Simplify stats-date-picker date handling

Removed unnecessary intermediate momentDate variable and unused moment prop
since momentSiteZone can handle both moment objects and date strings directly.

kangzj committed:
Stats: Add TODO comment for rangeOfPeriod timezone handling

kangzj committed:
Stats: Add essential GMT offset date comparison tests

Added 4 focused tests for GMT offset handling to prevent regression
of the arrow navigation bug and ensure date comparison logic works correctly.

Essential test coverage:
1. isSameOrAfter comparisons work correctly with GMT offset
2. Midnight edge case (23:59 vs 00:00) handled properly
3. Consistency between current time path and string parsing path
4. Fractional GMT offset (e.g., IST +5:30) works correctly

These tests directly address the bug being fixed where date comparisons
failed with GMT offset due to inconsistent handling between the "today"
path and string parsing.

All 48 tests passing (added 4 essential tests).

kangzj committed:
Stats: Streamline useMomentSiteZone tests to essential coverage

Reduced test suite from 48 to 14 essential tests by removing redundant scenarios while maintaining critical coverage for timezone handling and bug prevention.

kangzj commented on client/my-sites/stats/hooks/use-moment-site-zone.ts:
```diff
@@ -29,11 +29,20 @@ export const getMomentSiteZone = createSelector(
-					);
+					// Get current time in site timezone and format as date string, then create
+					// a moment with the site's UTC offset applied for consistent comparisons.
+					const todayInSiteZone = moment().utcOffset( gmtOffset ).format( dateFormat );
```
Good catch here!

kangzj committed:
Stats: Remove unused useMomentInSite import from realtime chart

The realtime chart uses browser time for display purposes, not site timezone.

kangzj commented on client/my-sites/stats/hooks/use-moment-site-zone.ts:
```diff
@@ -6,10 +6,24 @@ import getSiteGmtOffset from 'calypso/state/selectors/get-site-gmt-offset';
+ */
 export const getMomentSiteZone = createSelector(
-	( state: object, siteId: number | null, dateFormat = DATE_FORMAT ) => {
+	( state: object, siteId: number | null ) => {
```
There should never been dateFormat - it was an oversight adding this in.

kangzj commented on client/my-sites/stats/pages/realtime/chart.tsx:
```diff
@@ -27,16 +26,15 @@ const UPDATE_INTERVAL_IN_SECONDS = 5;
 const MINUTE_DATA_LENGTH = 30;
 
 const RealtimeChart = ( { siteId }: { siteId: number } ) => {
-	const momentInSite = useMomentInSite( siteId );
```
Reverting those changes, as there are a few things that needs to be adjusted if we want to use site timezone here.

kangzj commented on client/my-sites/stats/site.jsx:
```diff
@@ -262,10 +262,13 @@ function StatsBody( { siteId, chartTab = 'views', date, context, isInternal, ...
-		if ( currentChartStart && momentInSite( chartStart ).isBefore( currentChartStart ) ) {
+		if (
+			currentChartStart &&
+			momentInSite( chartStart ).isBefore( momentInSite( currentChartStart ) )
```
When comparing, we should covert string to moment in site timezone

kangzj commented on client/my-sites/stats/hooks/test/use-moment-site-zone.test.ts:
```diff
@@ -0,0 +1,248 @@
+/**
```
Added in tests

kangzj commented:
Thanks for fixing this @Nikschavan - that was a nice catch! I removed some legacy and unnecessary code and also added some test 😃
